Otter Grade: Added Progress Monitoring

sean-morris opened the pull request:
- When an instance of ``multiprocessing.Queue`` is passed to grade/__init_.py::main, progress messages are written to the ``multiprocessing.Queue`


Closes #827

chrispyles commented on otter/cli.py:
```diff
@@ -3,6 +3,7 @@
 import click
 import functools
 import logging
+import asyncio
```
nit: keep imports sorted

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -2,12 +2,14 @@
 
 import os
 import re
+import queue
```
nit: keep imports sorted

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -2,12 +2,14 @@
 
 from .containers import launch_containers
 from .utils import (
+    monitor_progress,
```
nit: keep imports sorted

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -43,6 +45,8 @@ def main(
     no_network: bool = False,
     debug: bool = False,
+    result_queue: queue.Queue = None,
+    print_progress: bool = False
```
nit: add trailing comma

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -117,7 +123,8 @@ def main(
     pdf_dir = os.path.join(output_dir, "submission_pdfs") if pdfs else None
 
-    grade_dfs = launch_containers(
+    monitor_progress(result_queue, print_progress, "Launching Grading Process")
```
Again, please make use of the existing logging infrastructure by logging to `LOGGER`. If you want logs to also be written to the queue, I'd suggest creating a new logging handler that does this.

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -132,10 +139,14 @@ def main(
             "debug": debug,
         }),
+        result_queue = result_queue,
+        print_progress = print_progress
```
nit: add trailing comma

chrispyles commented on otter/grade/containers.py:
```diff
@@ -7,18 +7,23 @@
 import zipfile
+import queue
+import asyncio
+import nest_asyncio
```
nit: import ordering

chrispyles commented on otter/grade/containers.py:
```diff
@@ -7,18 +7,23 @@
 from typing import List, Optional
 
-from .utils import OTTER_DOCKER_IMAGE_NAME, merge_scores_to_df, TimeoutException
+from .utils import OTTER_DOCKER_IMAGE_NAME, merge_scores_to_df, TimeoutException, monitor_progress
```
nit: import ordering

chrispyles commented on otter/grade/containers.py:
```diff
@@ -104,30 +121,39 @@ def launch_containers(
+            grade_submission(
+                submission_path=subm_path,
+                image=image,
+                # config=config,
```
can you please remove this? ik you didn't add it but not sure why this is still here

sean-morris commented:
@chrispyles I needed the grade/__init__.py::main function to be async so that I could call it directly from my application passing an instance of ``queue.Queue``.  This is instead of calling ``otter grade`` using the cli command and the python subprocess library -- which never felt right to me.

I added the --print-progress flag so that progress could printed to the console when using CLI.

I am open to whatever here Chris.  A first set of moments.

chrispyles commented on otter/grade/containers.py:
```diff
@@ -104,30 +121,39 @@ def launch_containers(
-    # stop execution while containers are running
-    finished_futures = wait(futures)
-    scores = [f.result() for f in finished_futures[0]]
+        futures += [asyncio.wrap_future(pool.submit(
```
Why are you adding asyncio? I don't see what this adds that wasn't available just from the ThreadPoolExecutor?

chrispyles commented on otter/grade/utils.py:
```diff
@@ -3,6 +3,7 @@
 import os
 import pandas as pd
 import re
+import queue
```
nit: sort imports

chrispyles commented on test/test_grade/test_integration.py:
```diff
@@ -7,6 +7,10 @@
+import asyncio
+import queue
+import io
+import sys
```
nit: sort imports

chrispyles reviewed (changes requested):
This is a pretty big change. I think this should go into a major release (i.e. v6) so can you please rebase against the beta branch and PR to it?

chrispyles commented:
> This is instead of calling otter grade using the cli command and the python subprocess library -- which never felt right to me.

I'm not sure if this is a good idea -- by calling otter as a package directly from your server, if otter crashes, it could bring down your entire server. I think it's better to run otter in a separate process, which subprocess does. Alternatively, if you want to call otter programmaticaly but still in a separate process, I think the multiprocessing module can be used to spawn a separate process that won't take down your server if it crashes (but I haven't used this much so I'd double check that).

sean-morris commented on otter/cli.py:
```diff
@@ -141,13 +142,15 @@ def generate_cli(*args, **kwargs):
 @click.option("--no-kill", is_flag=True, help="Do not kill containers after grading")
 @click.option("--debug", is_flag=True, help="Run in debug mode (without ignoring errors thrown during execution)")
 @click.option("--prune", is_flag=True, help="Prune all of Otter's grading images")
+@click.option("--print-progress", is_flag=True, help="Progress remarks are written to console")
```
yes ok, so I am just going to remove print-progress

sean-morris commented on test/test_grade/test_integration.py:
```diff
@@ -401,3 +407,53 @@ def test_grade_summaries(mocked_launch_grade):
+
+@pytest.mark.slow
+@pytest.mark.docker
+def test_print_progress():
```
removed this with removal of print-progress flag

sean-morris commented:
> @chrispyles I needed the grade/**init**.py::main function to be async so that I could call it directly from my application passing an instance of `queue.Queue`. This is instead of calling `otter grade` using the cli command and the python subprocess library -- which never felt right to me.
> 
> I added the --print-progress flag so that progress could printed to the console when using CLI. I am open to whatever here Chris. A first set of moments.

@chrispyles Scratch all that.  I used your multiprocessing module idea and this greatly simplified just about everything! I also removed the --print-progress piece altogether.

chrispyles commented on otter/grade/utils.py:
```diff
@@ -151,3 +152,15 @@ def merge_scores_to_df(scores: List[GradingResults]) -> pd.DataFrame:
     return full_df
+
+
+def monitor_progress(queue: Queue, msg: str):
```
Since Otter already has logging, why not hook into that (and add your own logs as needed)? You could create a [logging handler](https://docs.python.org/3/library/logging.html#handler-objects) that handles logs by passing them to your `Queue` and do something similar to [this](https://github.com/ucbds-infra/otter-grader/blob/master/otter/utils.py#L268) in the `loggers` class to send any logs to that handler when it's set up.

sean-morris replied:
I pushed a commit that uses the logger. Take a look. I was not quite sure how to approach registering my handler with the loggers. Open to suggestions.

chrispyles commented:
> This is a pretty big change. I think this should go into a major release (i.e. v6) so can you please rebase against the beta branch and PR to it?

Also please don't forget to move this to the beta branch @sean-morris

sean-morris commented:
> > This is a pretty big change. I think this should go into a major release (i.e. v6) so can you please rebase against the beta branch and PR to it?
> 
> Also please don't forget to move this to the beta branch @sean-morris

In order to do this, we should rebase the master branch into the beta branch and then bring over my commit from feature branch? Lame here. It is a mix of commits that is slightly nerve-racking to merge.

chrispyles commented:
All of the commits in the master branch are already merged into the beta branch. you should just be able to rebase your branch onto beta, or checkout to the beta branch, create a new branch off of it, and cherry-pick your commits into it.

sean-morris commented:
> > > This is a pretty big change. I think this should go into a major release (i.e. v6) so can you please rebase against the beta branch and PR to it?
> > 
> > 
> > Also please don't forget to move this to the beta branch @sean-morris
> 
> In order to do this, we should rebase the master branch into the beta branch and then bring over my commit from feature branch? Lame here. It is a mix of commits that is slightly nerve-racking to merge. @chrispyles

let me take another look. I am screwed up about this -- I keep seeing commits that should already be in master showing up in my feedback branch commits when I rebase it

chrispyles commented on otter/utils.py:
```diff
@@ -364,6 +380,31 @@ def _get_logger(cls):
+
+        Args:
+            log_queue (multiprocessing.Queue): the queue this handler writes to
+        """
```
`__init__` docstring is not needed. Move the `Args` here into the class docstring.

chrispyles commented on otter/utils.py:
```diff
@@ -364,6 +380,31 @@ def _get_logger(cls):
+    `otter grade`
+
+    Args:
+        logging (Hanlder): the base class for this handler
```
Remove this -- the `Args` here should be the args to `__init__`.

chrispyles commented on otter/utils.py:
```diff
@@ -364,6 +380,31 @@ def _get_logger(cls):
+class QueueLoggingHandler(logging.Handler):
+    """The logging handler that writes INFO messages to the log_queue.
+    This handler is set up when the result_queue is passed to
+    `otter grade`
```
Change the top of the docstring to

```
"""
This logging handler writes log messages to a ``multiprocessing.Queue``.
```

The information about it being set up in otter grade is unnecessary since this could be used by other components.

chrispyles commented on otter/utils.py:
```diff
@@ -364,6 +380,31 @@ def _get_logger(cls):
+    Args:
+        logging (Hanlder): the base class for this handler
+    """
+    def __init__(self, log_queue: Queue):
```
nit: add blank line between class docstring and `__init__`

chrispyles commented on test/test_grade/test_integration.py:
```diff
@@ -244,6 +244,7 @@ def test_single_notebook_grade(mocked_launch_grade):
         "timeout": None,
         "network": True,
         "config": AutograderConfig(),
+        "result_queue": None
```
nit: add trailing comma

chrispyles commented on otter/utils.py:
```diff
@@ -290,9 +292,23 @@ class log level.
+    @classmethod
+    def add_queue_handler(cls, result_queue):
+        """
+        adds queue handler to all instances of the logger
```
nit: docstring should be formatted and reworded as 

```
"""
Set up a ``QueueLoggingHandler`` that sends all logged messages to the provided 
``multiprocessing.Queue``.

Args:
    result_queue (``multiprocessing.Queue``): the queue to write logs to
"""
```

chrispyles replied:
bump

chrispyles commented on test/test_grade/test_integration.py:
```diff
@@ -401,3 +402,24 @@ def test_grade_summaries(mocked_launch_grade):
+        output_dir = "test/",
+        autograder = AG_ZIP_PATH,
+        summaries = False,
+        result_queue = test_queue
```
nit: add trailing comma

chrispyles replied:
bump

chrispyles commented on test/test_grade/test_integration.py:
```diff
@@ -401,3 +402,24 @@ def test_grade_summaries(mocked_launch_grade):
+    )
+    with open(FILE_MANAGER.get_path("queue/passall_messages.txt"), "r") as f:
+        for line in f:
+            assert test_queue.get() == line.strip()
```
can you add a check that the queue is empty at the end of the test

chrispyles replied:
bump

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -132,10 +138,12 @@ def main(
 
     LOGGER.info("Combining grades and saving")
 
+
```
nit: remove this line

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -43,6 +44,7 @@ def main(
     timeout: bool = None,
     no_network: bool = False,
     debug: bool = False,
+    result_queue: Queue = None,
```
type annotation should be `Optional[Queue]`

sean-morris replied:
I got this change in

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -71,6 +73,7 @@ def main(
         timeout (``int``): an execution timeout in seconds for each container
         no_network (``bool``): whether to disable networking in the containers
         debug (``bool``): whether to run autograding in debug mode
+        result_queue (``multiprocessing.Queue``): the queue to store progress messages
```
type should be ``multiprocessing.Queue | None``

chrispyles commented on otter/grade/containers.py:
```diff
@@ -111,13 +114,17 @@ def launch_containers(
-            # config=config,
-            **kwargs,
+            result_queue=result_queue,
+            **kwargs
```
nit: keep trailing comma

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -103,6 +106,9 @@ def main(
         raise ValueError(f"Invalid submission extension specified: {ext}")
 
+    if result_queue:
+        loggers.add_queue_handler(result_queue)
```
Since the logging handler is created here, `result_queue` doesn't need to be passed anywhere in containers.py, right?

chrispyles commented on otter/grade/__init__.py:
```diff
@@ -103,6 +106,9 @@ def main(
         raise ValueError(f"Invalid submission extension specified: {ext}")
 
+    if result_queue:
+        loggers.add_queue_handler(result_queue)
```
The queue logging handler should be removed when this function exits, in case the same python process calls this function multiple times. I think the best way to accomplish this would be to wrap the function body in a `try` block that removes the logging handler in a `finally` block.

sean-morris replied:
I should have been clearer @chrispyles -- so you are cool with how I bring the queue handler into the logger? I will remove all result_queue arguments, and add the try ... finally to remove the queue handler. I got everything into the beta branch. It got confusing because I did not realize you had changed some of my earlier implementations with later commits -- which is fine. I ended up cherry-picking my two commits from the feedback branch and putting them on the beta branch -- It was easier for me to see what was happening and integrate your changes to the files I had touched, and not screw up any of the other changes you made to other files!

I think we are close on all this. Thanks for your feedback.

chrispyles replied:
> so you are cool with how I bring the queue handler into the logger

Yes, this approach LGTM.

sean-morris replied:
> The queue logging handler should be removed when this function exits, in case the same python process calls this function multiple times. I think the best way to accomplish this would be to wrap the function body in a `try` block that removes the logging handler in a `finally` block.

@chrispyles I think you meant to put the try...finally block around the launch_containers function call in grade/__init__.py. if not tell me. This makes sense though in general.

chrispyles replied:
The try/finally should encompass everything from line 111 to at least line 145, assuming that nothing after line 145 generates logs you care about. The reason why is because if an error is thrown before launch_containers is called, we still want to ensure that the logging handler is removed.

chrispyles commented on otter/grade/containers.py:
```diff
@@ -236,9 +238,12 @@ def kill_container():
 
+    except TimeoutException as te:
+        scores = GradingResults.without_results(te)
+        LOGGER.info(f"Notebook Grading Timeout Error: {nb_basename}")
```
Should this be logged as an error (i.e. with `LOGGER.error`)?

chrispyles commented on otter/grade/containers.py:
```diff
@@ -236,9 +238,12 @@ def kill_container():
     except Exception as e:
         scores = GradingResults.without_results(e)
-
+        LOGGER.info(f"Notebook Grading Error: {nb_basename}")
```
Should this be logged as an error (i.e. with `LOGGER.error`)?

sean-morris commented:
You are running ahead of me! I just changed the merge branch to be beta not master. Hope that was OK.

chrispyles commented on otter/utils.py:
```diff
@@ -280,9 +282,33 @@ class log level.
             logger.addHandler(cls._socket_handler)
+        if cls._queue_handler:
+            logger.addHandler(cls._queue_handler)
+            logger.setLevel(logging.INFO)
```
This shouldn't alter the logging level since it's controller by a CLI flag and this will disrupt it. If your code is calling Otter via the CLI, use the `-v` flag to set the log level to INFO, or if it's calling otter as a library, have your code set the level of with `loggers.set_level` as in https://github.com/ucbds-infra/otter-grader/blob/master/otter/cli.py#L31.

sean-morris replied:
OK yes, I didn't like that moment. I am with you

chrispyles commented on otter/utils.py:
```diff
@@ -354,6 +380,25 @@ def _get_logger(cls):
+    """The logging handler that writes INFO messages to the log_queue.
+
+    Args:
+        log_queue (multiprocessing.Queue): the queue this handler writes to
```
nit: wrap argument type in double-backticks (i.e. ``` ``multiprocessing.Queue`` ```).

chrispyles commented:
> You are running ahead of me! I just changed the merge branch to be beta not master. Hope that was OK.

no worries!

sean-morris commented:
@chrispyles when I make these requested changes -- should I be rebasing/force pushing to keep it all as one commit? What is the best practice here?

chrispyles commented:
In general force pushing is a bad idea. You don't need to keep everything in a single commit, it's fine to have multiple. And now that the branch has all of the beta commits in it you shouldn't need to rebase again.

chrispyles commented on otter/grade/containers.py:
```diff
@@ -236,9 +238,12 @@ def kill_container():
     except Exception as e:
         scores = GradingResults.without_results(e)
-
+        LOGGER.error(f"Notebook Grading Error: {nb_basename}")
```
nit: add back the blank line after this one

chrispyles commented on otter/utils.py:
```diff
@@ -280,9 +282,35 @@ class log level.
+        """
+        Remove instance of ``QueueLoggingHandler`` from loggers
+        """
+        for name in cls._instances:
```
nit: this loop can be simplified as

```
for logger in cls._instances.values():
    logger.removeHandler(cls._queue_handler)
```

chrispyles reviewed (changes requested):
Two more nits and also please add an entry to the changelog for this, and then this is good to go. Thanks for going through all these revisions!
